src/tcache.c

```diff
@@ -121,6 +121,80 @@ tcache_gc_dalloc_postponed_event_wait(tsd_t *tsd) {
 	return TE_MIN_START_WAIT;
 }
 
+static inline void
+tcache_nfill_small_ctl_init(tcache_slow_t *tcache_slow, szind_t szind) {
+	cache_bin_fill_ctl_t *ctl =
+	    &tcache_slow->bin_fill_ctl_do_not_access_directly[szind];
+	ctl->base = 1;
+	ctl->offset = 0;
+}
+
+/*
+ * The number of items to be filled at a time for a given small bin is
+ * calculated by (ncached_max >> lg_fill_div).
+ * The actual ctl struct consists of two fields, i.e. base and offset,
+ * and the difference between the two(base - offset) is the final lg_fill_div.
+ * The base is adjusted during GC based on the traffic within a period of time,
+ * while the offset is updated in real time to handle the immediate traffic.
+ */
+static inline uint8_t
+tcache_nfill_small_lg_div_get(tcache_slow_t *tcache_slow, szind_t szind) {
+	cache_bin_fill_ctl_t *ctl =
+	    &tcache_slow->bin_fill_ctl_do_not_access_directly[szind];
+	assert(ctl->base > ctl->offset);
```

Comment:
Add a getter function, e.g. 

```
cache_bin_fill_ctl_get(tcache_slow) {
    cache_bin_fill_ctl_t *ctl = &tcache_slow->bin_fill_ctl_do_not_access_directly[szind];
    assert(ctl->base > ctl->offset);
    return ctl;
}
```
When we have the do_not_access_directly fields, we should limit where they are referenced. Typically only in getter, setter and init functions. You also don't have to repeat the asserts in the callers.